video/decode/vd_lavc.c

```diff
@@ -475,16 +479,25 @@ static void select_and_set_hwdec(struct mp_filter *vd)
 
         bool hwdec_requested = !bstr_equals0(opt, "no");
         bool hwdec_auto_all = bstr_equals0(opt, "auto") ||
-                            bstr_equals0(opt, "yes") ||
                             bstr_equals0(opt, "");
         bool hwdec_auto_safe = bstr_equals0(opt, "auto-safe") ||
-                            bstr_equals0(opt, "auto-copy-safe");
+                            bstr_equals0(opt, "auto-copy-safe") ||
+                            bstr_equals0(opt, "yes");
         bool hwdec_auto_copy = bstr_equals0(opt, "auto-copy") ||
                             bstr_equals0(opt, "auto-copy-safe");
         bool hwdec_auto = hwdec_auto_all || hwdec_auto_copy || hwdec_auto_safe;
 
         if (!hwdec_requested) {
             MP_VERBOSE(vd, "No hardware decoding requested.\n");
+            /*
+             * Reset attempted hwdecs so that if the hwdec list is reconfigured
+             * we attempts all of them from the beginning. The most practical
+             * reason for this is that ctrl+h toggles between `no` and
+             * `auto-safe`, and we want to reevaluate from a clean slate each time.
+             */
+            TA_FREEP(&ctx->attempted_hwdecs);
+            ctx->num_attempted_hwdecs = 0;
+            ctx->hwdec_notified = false;
```

Comment:
this probably needs to be changed in a way that vd_lavc is notified if the `hwdec` option changes

Reply:
Yes. I've moved this logic into `reinit`, which is called whenever the config option changes value. Should address @arm64-v9a 's report.